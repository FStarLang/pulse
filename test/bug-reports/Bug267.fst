module Bug267

#lang-pulse
open Pulse
open FStar.Mul

(* Complains that 'v is ghost, good. *)
[@@expect_failure [228]]
fn value_of (#a:Type) (r:ref a)
  requires pts_to r 'v
  returns v:a
  ensures pts_to r 'v ** pure (v == 'v)
{
    'v
}

fn add (r : ref int) (v : int)
  requires pts_to r 'v0
  ensures  pts_to r (v + 'v0)
{
  let v0 = !r;
  r := v0 + v
}

(* Complains in !i: expected int got stt .... Could be better. *)
[@@expect_failure [12]]
fn four_fail ()
  requires emp
  returns i:int
  ensures pure (i == 4)
{
    let mut i = 1;
    add i (!i);
    add i (!i);
    !i
}

[@@expect_failure [19]]
fn multiply_by_repeated_addition (x y:nat)
    requires emp
    returns z:nat
    ensures pure (z == x * y)
{
    let mut ctr = 0;
    let mut acc : int = 0;
    while (
        let c = !ctr;
        (c < x)
    )
    invariant b.
    exists* (c a : int).
        pts_to ctr c **
        Pulse.Lib.Reference.pts_to acc a **
        pure (//0 <= c /\
              c <= x /\
              a == (c * y) /\
              b == (c < x))
    {
        let a = !acc;
        acc := a + y;
        let c = !ctr;
        ctr := c + 1;
    };
    !acc
}

[@@expect_failure [12]]
fn foo (n: nat) requires pure (n >= 42) ensures emp {
  assert (n >= 1);
}

