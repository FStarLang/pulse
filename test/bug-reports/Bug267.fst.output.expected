* Info at Bug267.fst(14,4-14,6):
  - Expected failure:
  - Tactic failed
  - Expected a total term, but this term has Ghost effect

* Info at Bug267.fst(62,4-62,5):
  - Expected failure:
  - Could not prove equality between computed type Prims.int
    and expected type Prims.nat
  - Assertion failed
  - The SMT solver could not prove the query. Use --query_stats for more
    details.
  - Also see: Prims.fst(161,28-183,79)
  - See also Bug267.fst(62,4-62,8)

* Info at Bug267.fst(67,13-67,14):
  - Expected failure:
  - Expected type Pulse.Lib.Core.slprop but n >= 1 has type Prims.bool

